src/vm/crossloaderallocatorhash.inl

```diff
@@ -0,0 +1,1216 @@
+
+        if (usedEntries == entriesInArrayTotal)
+        {
+            // There isn't free space. Build a new, bigger array with the existing data 
+            DWORD newSize;
+            if (usedEntries < 8)
+                newSize = usedEntries + 1; // Grow very slowly initially. The cost of allocation/copy is cheap, and this holds very tight on memory usage
+            else
+                newSize = usedEntries * 2;
+
+            if (newSize < usedEntries)
+                COMPlusThrow(kOverflowException);
+
+            // Allocate the new array.
+            I1ARRAYREF newKeyValueStore = (I1ARRAYREF)AllocatePrimitiveArray(ELEMENT_TYPE_I1, newSize*sizeof(TValue) + sizeof(TKey), FALSE);
+
+            // Since, AllocatePrimitiveArray may have triggered a GC, recapture all data pointers from GC objects
+            void* pStartOfNewArray = newKeyValueStore->GetDirectPointerToNonObjectElements();
+            void* pStartOfOldArray = ((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements();
+
+            memcpyNoGCRefs(pStartOfNewArray, pStartOfOldArray, ((I1ARRAYREF)*pKeyValueStore)->GetNumComponents());
+
+            *pKeyValueStore = (OBJECTREF)newKeyValueStore;
+            updatedKeyValueStore = true;
+
+            entriesInArrayTotal = newSize;
+        }
+
+        // There is free space. Append on the end
+        TValue* pStartOfValuesData = (TValue*)(((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements() + sizeof(TKey));
+        SetUsedEntries(pStartOfValuesData, entriesInArrayTotal, usedEntries + 1);
+        pStartOfValuesData[usedEntries] = value;
+    }
+
+    return updatedKeyValueStore;
+}
+#endif //!DACCESS_COMPILE
+
+template <class TKey_, class TValue_>
+/*static*/ TKey_ NoRemoveDefaultCrossLoaderAllocatorHashTraits<TKey_, TValue_>::ReadKeyFromKeyValueStore(OBJECTREF *pKeyValueStore)
+{
+    WRAPPER_NO_CONTRACT;
+
+    TKey* pKeyLoc = (TKey*)((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements();
+    return *pKeyLoc;
+}
+
+template <class TKey_, class TValue_>
+template <class Visitor>
+/*static*/ bool NoRemoveDefaultCrossLoaderAllocatorHashTraits<TKey_, TValue_>::VisitKeyValueStore(OBJECTREF *pLoaderAllocatorRef, OBJECTREF *pKeyValueStore, Visitor &visitor)
+{
+    WRAPPER_NO_CONTRACT;
+
+    DWORD entriesInArrayTotal;
+    DWORD usedEntries = ComputeUsedEntries(pKeyValueStore, &entriesInArrayTotal);
+
+    for (DWORD index = 0; index < usedEntries; ++index)
+    {
+        // Capture pKeyLoc and pStartOfValuesData inside of loop, as we aren't protecting these pointers into the GC heap, so they
+        // are not permitted to live across the call to visitor (in case visitor triggers a GC)
+        TKey* pKeyLoc = (TKey*)((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements();
+        TValue* pStartOfValuesData = (TValue*)(((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements() + sizeof(TKey));
```

Comment:
Do these need to use `VolatileLoadWithoutBarrier` to prevent the compiler from hoisting them out of the loop?

Reply:
No, since pKeyValueStore is on the heap (in C++ semantics) if visitor may trigger a GC or poll for a GC, the GC triggering path will take a lock which is considered to be a sufficient instruction to the compiler that it can't hoist the load. If it won't trigger a GC, then the hoist would be valid.

Reply:
Gotcha